File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    answer = n
    graph = [[] for _ in range(n+1)]
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)
    
    for cuta, cutb in wires:
        visited = [False] * (n +1)
        q = deque([1])
        visited[1] = True
        cnt= 1
        while q:
            cur = q.popleft()
            for nxt in graph[cur]:
                
                if (cur == cuta and nxt == cutb) or (cur == cutb and nxt == cuta):
                    continue
                
                if not visited[nxt]:
                    visited[nxt] = True
                    q.append(nxt)
                    cnt+=1
        other = n-cnt
        answer = min(abs(other-cnt), answer)
    
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int n, vector<vector<int>> wires) {
    int answer = n;
    vector<vector<int>> graph(n+1);
    for(auto a:wires){
        graph[a[0]].push_back(a[1]);
        graph[a[1]].push_back(a[0]);
    }


    for(auto a:wires){
        int cut_a=a[0];
        int cut_b=a[1];
        queue<int> q;
        vector<bool> visited(n+1);
        q.push(1);
        visited[1]=true;
        int cnt=0;
        
        while(!q.empty()){
            int curr = q.front();
            q.pop();
            cnt++;

            for(auto next: graph[curr]){
                if((curr==cut_a && next==cut_b)|| (curr==cut_b && next==cut_a)){
                    continue;
                }
                if(!visited[next]){
                    q.push(next);
                    visited[next]=true;
                }
            }
        }
        int other = n-cnt;
        answer = min(answer, abs(cnt-other));
    }
    return answer;
}